### PR TITLE
shared/tls/sni: Dispatcher, a forwarder beneath tls.Listener

### PR DESCRIPTION
**Pull request by amery**

### CODE REVIEW

File: shared/tls/sni/chi.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+func (c *Conn) Read(b []byte) (int, error) {
+	return c.Reader.Read(b)
+}
+
+// PeekClientHelloInfo extracts the ClientHelloInfo from a connection
+// still allowing a future handler have complete untouched access to
+// the stream
+func PeekClientHelloInfo(ctx context.Context, conn net.Conn) (
+	*tls.ClientHelloInfo, net.Conn, error) {
```
Comment (karasz) [inline shared/tls/sni/chi.go:26]: maybe we should try this pattern when breaking long function definitions:
```go
func PeekClientHelloInfo(ctx context.Context, conn net.Conn) (
     *tls.CientHelloInfo, net.Conn, error,
) {
```

**Review by karasz — APPROVED** [review]

**Comment by amery** [comment]
amended

File: shared/tls/sni/chi.go (hunk 1 above)
Comment (amery) [reply shared/tls/sni/chi.go]: @karasz what about now?

Comment (karasz) [reply shared/tls/sni/chi.go]: +1

**Comment by amery** [comment]
`shared/v0.3.1` tagged

**Pull request merged**